File: src/DIY/AbstractFactory/AbstractFactory.java
package DIY.AbstractFactory;

public class AbstractFactory {
    public static void main(String[] args) {
        APIFactory weatherFactory = new WeatherAPIFactory();
        APIRequest weatherRequest = weatherFactory.createRequest();
        APIResponse weatherResponse  = weatherFactory.createResponse();

        weatherRequest.sendRequest();
        weatherResponse.sendResponse();

        System.out.println("-----");
        APIFactory mapsFactory = new MapsAPIFactory();
        APIRequest mapsRequest = mapsFactory.createRequest();
        APIResponse mapsResponse  = mapsFactory.createResponse();

        mapsRequest.sendRequest();
        mapsResponse.sendResponse();
        
    }

}

interface APIFactory{
    APIRequest createRequest();
    APIResponse createResponse();
    
}

interface APIRequest{
    void sendRequest();
    
}

interface APIResponse{
    void sendResponse();

}

class WeatherRequest implements APIRequest{

    @Override
    public void sendRequest() {
        System.out.println("Sending a request to WeatherAPI...");
    }

}

class WeatherResponse implements APIResponse{

    @Override
    public void sendResponse() {
        System.out.println("Fetching weather data from WeatherAPI...");
    }

}

class MapRequest implements APIRequest{

    @Override
    public void sendRequest() {
        System.out.println("Sending a request to MapsAPI...");
    }

}
class MapResponse implements APIResponse{

    @Override
    public void sendResponse() {
        System.out.println("Fetching latest GPS data from MapsAPI...");
    }
    
}

class WeatherAPIFactory implements APIFactory{

    @Override
    public APIRequest createRequest() {
        return new WeatherRequest();
    }

    @Override
    public APIResponse createResponse() {
        return new WeatherResponse();
    }

}

class MapsAPIFactory implements APIFactory{

    @Override
    public APIRequest createRequest() {
        return new MapRequest();
    }

    @Override
    public APIResponse createResponse() {
        return new MapResponse();
    }

}

File: src/DIY/AbstractFactory/AbstractFactory.ts
interface APIFactory {
    createRequest(): APIRequest;
    createResponse(): APIResponse;
}

interface APIRequest {
    sendRequest(): void;
}

interface APIResponse {
    sendResponse(): void;
}

class WeatherRequest implements APIRequest {
    sendRequest(): void {
        console.log("Sending a request to WeatherAPI...");
    }
}

class WeatherResponse implements APIResponse {
    sendResponse(): void {
        console.log("Fetching weather data from WeatherAPI...");
    }
}

class MapRequest implements APIRequest {
    sendRequest(): void {
        console.log("Sending a request to MapsAPI...");
    }
}

class MapResponse implements APIResponse {
    sendResponse(): void {
        console.log("Fetching latest GPS data from MapsAPI...");
    }
}

class WeatherAPIFactory implements APIFactory {
    createRequest(): APIRequest {
        return new WeatherRequest();
    }

    createResponse(): APIResponse {
        return new WeatherResponse();
    }
}

class MapsAPIFactory implements APIFactory {
    createRequest(): APIRequest {
        return new MapRequest();
    }

    createResponse(): APIResponse {
        return new MapResponse();
    }
}

function main(): void {
    const weatherFactory: APIFactory = new WeatherAPIFactory();
    const weatherRequest: APIRequest = weatherFactory.createRequest();
    const weatherResponse: APIResponse = weatherFactory.createResponse();

    weatherRequest.sendRequest();
    weatherResponse.sendResponse();

    console.log("-----");

    const mapsFactory: APIFactory = new MapsAPIFactory();
    const mapsRequest: APIRequest = mapsFactory.createRequest();
    const mapsResponse: APIResponse = mapsFactory.createResponse();

    mapsRequest.sendRequest();
    mapsResponse.sendResponse();
}

main();
